linkerd/dns/src/lib.rs

```diff
@@ -3,45 +3,60 @@
 mod refine;
 
 pub use self::refine::{MakeRefine, Refine};
+use async_trait::async_trait;
 pub use linkerd2_dns_name::{InvalidName, Name, Suffix};
 use std::future::Future;
 use std::pin::Pin;
 use std::{fmt, net};
 use tokio::sync::{mpsc, oneshot};
+use tokio::time::Instant;
 use tracing::{info_span, trace, Span};
 use tracing_futures::Instrument;
 pub use trust_dns_resolver::config::ResolverOpts;
 pub use trust_dns_resolver::error::{ResolveError, ResolveErrorKind};
-use trust_dns_resolver::lookup_ip::LookupIp;
+pub use trust_dns_resolver::lookup_ip::LookupIp;
 use trust_dns_resolver::{config::ResolverConfig, system_conf, AsyncResolver};
 
+#[derive(Clone, Debug)]
+pub struct ResolveResponse {
+    pub ips: Vec<net::IpAddr>,
+    pub valid_until: Instant,
+}
+
+struct ResolveRequest {
+    name: Name,
+    result_tx: oneshot::Sender<Result<LookupIp, ResolveError>>,
+    span: tracing::Span,
+}
+
+#[async_trait]
+pub trait Resolver: Clone {
+    async fn lookup_ip(&self, name: Name, span: Span) -> Result<LookupIp, Error>;
+    async fn resolve_ips(&self, name: &Name) -> Result<ResolveResponse, Error>;
+    /// Creates a refining service.
+    fn into_make_refine(self) -> MakeRefine<Self>;
+}
+
 #[derive(Clone)]
-pub struct Resolver {
+pub struct DnsResolver {
     tx: mpsc::UnboundedSender<ResolveRequest>,
 }
 
 pub trait ConfigureResolver {
     fn configure_resolver(&self, _: &mut ResolverOpts);
 }
 
-#[derive(Debug)]
+#[derive(Debug, Clone)]
 pub enum Error {
-    NoAddressesFound,
+    NoAddressesFound(Instant, bool),
```

Comment:
please use named fields here. It's not at all obvious what the intent is from looking at the type.